classes/local/client/s3_client.php

```diff
@@ -39,64 +39,160 @@
  */
 class s3_client {
 
+    /**
+     * Plugin config.
+     * @var false|mixed|object|string
+     */
+    private $config;
+
+    /**
+     * Client.
+     * @var \Aws\AwsClientInterface|null
+     */
+    private $client = null;
+
     /**
      * Constructor for S3 client class.
-     * Makes relevant config available and bootstraps
-     * AWS S3 client.
+     *
+     * Makes relevant config available and bootstraps AWS S3 client.
      *
      * @return void
      */
     public function __construct() {
         $this->config = get_config('tool_s3logs');
-        $this->s3region = $this->config->s3region;
-        $this->keyid = $this->config->keyid;
-        $this->secretkey = $this->config->secretkey;
-        $this->bucket = $this->config->bucket;
-        $this->client = $this->get_s3_client();
+        $this->set_client();
     }
 
     /**
-     * Create AWS S3 client.
-     *
-     * @return client $s3client S3 client.
+     * Check if the client is functional.
+     * @return bool
+     */
+    private function is_functional(): bool {
+        return isset($this->client);
+    }
+
+    /**
+     * Check if the client configured properly.
+     * @return bool
      */
-    private function get_s3_client() {
-        $settings = array(
-            'region' => $this->s3region,
-            'version' => 'latest'
-        );
-        $usesdkcreds = $this->config->usesdkcreds;
-        if (!$usesdkcreds) {
-            $settings['credentials'] = array('key' => $this->keyid, 'secret' => $this->secretkey);
+    private function is_configured(): bool {
+        if (empty($this->config->bucket)) {
+            return false;
+        }
+
+        if (empty($this->config->s3region)) {
+            return false;
```

Comment:
Just a tiny one suggestion to make it simpler:

```suggestion
        if (empty($this->config->bucket) || empty($this->config->s3region)) {
            return false;
```

Reply:
Applied the fix.